### PR TITLE
Added reentrant locks module to stdlib

### PR DESCRIPTION
**Pull request by vegansk**
On windows, CRITICAL_SESSION is reentrant. Posix mutexes is non-reentrant by default. This module introduces reentrant locks with the same behaviour on windows and posix systems.

### CODE REVIEW

**Comment by vegansk** [comment]
The discussion is here: http://irclogs.nim-lang.org/17-02-2016.html (from 09:32:59)

File: lib/core/rlocks.nim [hunk 1]
```diff
@@ -0,0 +1,50 @@
+#
+#
+#            Nim's Runtime Library
+#        (c) Copyright 2015 Andreas Rumpf
```
Comment (Araq) [inline lib/core/rlocks.nim]: FYI using your own name here is perfectly fine ;-)

File: lib/core/rlocks.nim [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+proc acquire*(lock: var RLock) =
+  ## Acquires the given lock.
+  acquireSys(lock)
+
+proc release*(lock: var RLock) =
+  ## Releases the given lock.
+  releaseSys(lock)
+
+template withRLock*(lock: var RLock, code: stmt): stmt =
```
Comment (Araq) [inline lib/core/rlocks.nim]: Use 'untyped' instead of 'stmt'!

**Comment by vegansk** [comment]
@Araq, done :-)

**Comment by Araq** [comment]
Ok, now please update ``news.txt`` and ensure it ends up in the generated documentation.

**Comment by vegansk** [comment]
@Araq, done. And one more thing. ``syslocks`` module had a doc comment and after including it, this comment was mixed with ``locks`` and ``rlocks`` doc comments. So, I replaced it with the regular comment.

**Pull request merged**